v10.12: add documentation for server pre-shared secrets (#8319)
mattermost/docs: source/deployment-guide/server/pre-authentication-secrets.rst

      #    # ...
      # }
      # location ~ ^/(signup|login)/[A-Za-z0-9]+/complete$ {
      #    # ...
      # }

      location / {
          # Check if X-Mattermost-Preauth-Secret header matches expected value
          if ($http_x_mattermost_preauth_secret != $expected_secret) {
              add_header X-Reject-Reason pre-auth always;
              add_header Cache-Control "no-store" always;
              return 403 "Forbidden: Invalid pre-authentication secret";
          }

          # ...
      }
  }

Replace ``your-secure-pre-auth-secret-here`` with a strong, unique secret that will be configured in your mobile and desktop applications. Store this secret securely and rotate it regularly as part of your security practices.

Desktop application security considerations
-------------------------------------------

The Mattermost desktop application stores pre-authentication secrets using Electron's ``safeStorage`` API, which integrates with the operating system's secure credential storage:

- **Windows**: Windows Credential Manager
- **macOS**: Keychain Access
- **Linux**: Secret Service API (requires kwallet, gnome-libsecret, or compatible backend)

.. warning::

>   On Linux systems where no secure credential storage is available, the pre-authentication secret may be stored in **plain text**. This occurs when:

Reviewer: @fmartingr - Are you open to incorporating @enzowritescode's feedback in this PR?

>   On Linux systems where no secure credential storage is available, the pre-authentication secret may be stored in **plain text**. This occurs when:

Reviewer 2: I'll can add that!

Pre-authentication secrets
==========================

For mobile and desktop applications only, Mattermost deployments can use a reverse proxy to validate pre-authentication secrets before allowing requests to reach the Mattermost server. This adds an additional security layer by checking for the ``X-Mattermost-Preauth-Secret`` header.

> .. note::
> 
>   Pre-authentication secrets are only supported for mobile and desktop applications. Web browser clients do not support this feature.

Reviewer: ```suggestion
Pre-authentication secrets are only supported for mobile and desktop applications. Web browser clients don't support this feature.
```

.. important::

>   The ``/api/v4/notifications/ack`` endpoint should be whitelisted (allowed without pre-authentication secret validation) to ensure proper notification acknowledgment functionality for mobile applications.

Reviewer: ```suggestion
  We recommend whitelisting the ``/api/v4/notifications/ack`` endpoint (allowed without pre-authentication secret validation) to ensure proper notification acknowledgement functionality for mobile applications.
```

NGINX configuration example
---------------------------

Here's an example partial NGINX configuration that validates the pre-authentication secret header:

.. code-block:: text

  server {
      # ...

      # Define the expected pre-auth secret
      set $expected_secret "your-secure-pre-auth-secret-here";

      # Whitelist the notifications/ack endpoint (no pre-auth secret required)
      location = /api/v4/notifications/ack {
         # ...
      }

      location / {
          # Check if X-Mattermost-Preauth-Secret header matches expected value
          if ($http_x_mattermost_preauth_secret != $expected_secret) {
              return 403 "Forbidden: Invalid pre-authentication secret";
          }

          # ...
      }
  }

> .. important::
> 
>   Replace ``your-secure-pre-auth-secret-here`` with a strong, unique secret that will be configured in your mobile and desktop applications. Store this secret securely and rotate it regularly as part of your security practices.

Reviewer: ```suggestion
Replace ``your-secure-pre-auth-secret-here`` with a strong, unique secret that will be configured in your mobile and desktop applications. Store this secret securely and rotate it regularly as part of your security practices.
```
